design/coreos.md

```diff
@@ -0,0 +1,209 @@
+the maintainability perspective), we need an alternative way to start Ironic
+Python Agent on them. A canonical way to do is to to inject [Ignition
+configuration](https://coreos.github.io/ignition/examples/#start-services) that
+pulls a container and runs it with podman. [Example configuration
+template](https://github.com/dtantsur/ironic-agent-image/blob/main/ignition/ignition.json),
+[example service
+template](https://github.com/dtantsur/ironic-agent-image/blob/main/ignition/service).
+
+Note: Ignition is a CoreOS analog of cloud-init.
+
+We need to distinguish between two cases:
+
+1. PXE/iPXE booting
+2. ISO booting (with virtual media)
+
+CoreOS
+[publishes](https://getfedora.org/en/coreos/download?tab=metal_virtualized&stream=stable)
+both kinds of artifacts (and more), but the customization pattern is completely
+different.
+
+### PXE images
+
+CoreOS publishes 2 artifacts: a kernel, an initramfs and a root filesystem, all
+three are required for successful boot. The root filesystem and the Ignition
+configuration are specified as URLs via the kernel parameters.
+
+We will place the root filesystem and the generated Ignition configuration into
+the shared location where other HTTP artifacts are located. The Ironic's
+`pxe_append_params` will be updated with links to them.
+
+### ISO image
+
+We cannot modify kernel parameters on the ISO image without rebuilding it, thus
+we will use a different [customization
+procedure](https://coreos.github.io/coreos-installer/iso-embed-ignition/). On
+the Ironic start-up the downloaded ISO will be customized to include the
+generated Ignition configuration, including an Ironic Python Agent
+configuration file with the endpoint URLs (Ironic, Inspector, etc).
+
+### Enabling
+
+From the operator's perspective, the CoreOS IPA support will be enabled by
+setting a new environment variable `IRONIC_USE_COREOS_IPA=true`. By default the
+old RDO images will be used (this may change in the future as we gain more
+experience with CoreOS).
+
+It will also be possible to use a custom registry for the IPA container. By
+default `quay.io/metal3-io/ironic-agent` (to be created) will be used.
+
+### Invoking coreos-installer
+
+The default deployment procedure will work as before: by downloading an image
+and writing it to the disk.
+
+CoreOS installer support will be implemented using two new components:
+
+1. [New deploy interface](https://storyboard.openstack.org/#!/story/2008719)
+   that relies only on in-band [deploy
+   steps](https://docs.openstack.org/ironic/latest/admin/node-deployment.html).
+2. [Metal3-specific deploy
+   step](https://github.com/dtantsur/ironic-agent-image/blob/main/hardware_manager/ironic_coreos_install.py)
+   that will be injected into the IPA container and enabled explicitly via the
+   deployment API.
+
+From the user's perspective it will mean setting `Image.DiskFormat` to a new
+value `coreos` and leaving `Image.Url` empty.
```

Comment:
I'm not sure I like using disk format as the trigger for this behavior.

Where will the BMO get the URL for the image if the API field is empty?

Reply:
The CoreOS images have a root system built into them. There is no needed for having a separate image provided.

Reply:
I think the point from @dhellmann is if we don't require the user to specify `Image.Url`, the base image needs to be defined somewhere else (presumably another global variable?), then you can only use that single bootimage for all BMH resources, which is different from the current behavior of the BMH image API.

I'd be Ok with using DiskFormat for this, but IMO we should still require the BMH `Image.Url`, which would refer to the base coreos image (and this is also the trigger for BMH provisioning).

In this new flow, there would be an additional step though, the BMO would have to trigger the config-embed step (consuming the `Image.Url` with some way to locally cache and avoid duplicate downloads), then when completed the modified image can be added to the host in Ironic?

Reply:
The current pattern is to download the IPA image in a separate container. Do we want to change it? If so, I think we should introduce a new structure with IPA reference (it can be an ISO, or a kernel/initramfs pair, or kernel/initramfs/rootfs for CoreOS PXE artifacts).

Reply:
One of the key benefits of kubernetes' micro-services architecture is that if we create composable APIs we can place the responsibility for each part of a process in the "right" place. We don't have to put everything in the BMO, and in fact it is less complicated and more reusable if we decompose custom behaviors into the reusable parts and the custom parts.

For example, in this case we want to use a custom IPA image during inspection. Building and hosting the custom IPA image is one responsibility, and should be handled outside of the BMO. To do that, we need the host API to allow the user to provide image information for the IPA image, similar to the fields we have now for the target image. Then the CAPI provider could acquire the right IPA image (either a global image for the cluster or a host-specific image with static IP settings), and configure the host to use it instead of the default. When the host is provisioned, the host-specific image can be cleaned up. That approach doesn't require any CoreOS knowledge in the BMO.

Writing the OS to the host through an installer, instead of disk image provisioning, can be handled similarly. The CAPI provider that is provisioning the host will hold the knowledge of the need to use a custom installation approach. It knows what image to use, so presumably it can know how to run the installer. It needs appropriate API fields on the host to express its intent, to pass instructions through the BMO to Ironic and the custom IPA image. What those fields might be is less clear, but I see no need for them to be CoreOS-specific.

Reply:
> One of the key benefits of kubernetes' micro-services architecture is that if we create composable APIs we can place the responsibility for each part of a process in the "right" place

And one of the common misconceptions of kubernetes is that maintaining and deploying services is free ;)

> For example, in this case we want to use a custom IPA image during inspection.

This is not a part of this proposal yet, let's not try to solve everything at once. In this case the image is modified once per Ironic start-up, not per-host. We will indeed get to what you proposing, but when talking about static network configuration (which @hardys is looking in).

> That approach doesn't require any CoreOS knowledge in the BMO.

We still need to invoke deployment differently, it's not possible to avoid any knowledge. What you suggest is unnecessary complication, could you list at least one benefit (not modifying BMO is not a benefit)?

Reply:
On top of that, I'm worried about exposing `DeployInterface` and `DeploySteps` in public BMH API. These are very advanced Ironic features. I thought the idea was not to expose Ironic details to the users?

Reply:
> Doing it in a way that is not specific to CoreOS means it will be more reusable later, potentially more useful to other members of the community now, and not require members of the community who don't use CoreOS to understand it or directly support it.

A downside is that it will be (much) harder for non-OpenShift users to use the new functionality, and we already have at least one person using CoreOS without OpenShift.

> If the deploy step is defined to support running a user-specified command

This may work, but we'll lose a lot of flexibility. For example, right now the deploy step can accept custom ignition for the instance as a JSON. We won't be able to do it since there will be no way to write a temporary file for it. We'll have to use a URL which a user will have to provide (and host somewhere).

Similarly, we'll need some sort of templating to allow injecting the root device (calculated from the hints) into the provided command. This increases the overall complexity quite substantial, to a degree that it may be better to indeed expose ironic deploy step in the API.

Do we have an agreement that exposing deploy steps in BMH is better than changing BMO to explicitly support CoreOS?

Reply:
I don't think we want anything explicitly called "deploy steps" in the metal3 API. We may want to provide a way to pass arguments that are used by deploy steps. I don't know if that answers your question clearly enough.

Reply:
"Run this shell command" may not be sophisticated enough for our case, I'm afraid. I can make it work for the basic case, but I don't know if we'll be able to extend it if we need to. I will think about your proposal next week and we can discuss it on Tuesday.